[4855] The date/time displayed for scheduled publish is incorrect when approving a scheduled publish request

aachavarria opened the pull request:
Adding actionsBar to AwaitingApprovalDashlet.tsx
Moving selected logic to AwaitingApprovalDashlet.tsx
Using item.stateMap.submitted on PublishDialog.tsx to determinate when to show the approval checkbox

### Ticket reference or full description of what's in the PR

aachavarria committed:
Adding actions bar to AwaitingApprovalDashlet.tsx
Moving selected logic to AwaitingApprovalDashlet.tsx
Using item.stateMap.submitted on PublishDialog.tsx to determinate when to show the approval checkbox

aachavarria committed:
cleaning code

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4855

aachavarria committed:
adding submittedToStaging and submittedToLive to legacy item parse

aachavarria committed:
Updating dashlet to use lastScheduledDate

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4855

 Conflicts:
	ui/app/src/components/AwaitingApprovalDashlet/AwaitingApprovalDashlet.tsx
	ui/app/src/components/AwaitingApprovalDashletGrid/AwaitingApprovalDashletGridUI.tsx
	ui/app/src/utils/state.ts

aachavarria committed:
typo

aachavarria committed:
PublishDialog: Adding mixedPublishingTargets and Dates messages
AwaitingApprovalDashlet: adding clear option;

aachavarria committed:
Adding action bar to ApprovedScheduledDashlet.tsx

rart commented on ui/app/src/utils/itemActions.ts:
```diff
@@ -471,9 +471,9 @@ export const itemActionDispatcher = ({
   formatMessage;
-  clipboard;
+  clipboard?;
   onActionSuccess?: any;
```
Please fill in the correct types

rart commented on ui/app/src/modules/Content/Publish/PublishForm.tsx:
```diff
@@ -283,6 +281,14 @@ function PublishForm(props: PublishFormProps) {
+                defaultMessage="Items have mixed publishing date/time schedules."
+              />
+            </FormHelperText>
+          )}
```
I was thinking an alert-like message. 
Please send screenshot.

rart commented on ui/app/src/components/ApprovedScheduledDashlet/translations.ts:
```diff
@@ -0,0 +1,34 @@
+  reject: {
+    id: 'dashboardItemsScheduled.reject',
+    defaultMessage: 'Reject ({count})'
+  },
```
should be reject, schedule, publish

aachavarria committed:
updating approvePublish itemAction to send scheduling as null;
PublishDialog.tsx updates to support scheduling as null;

aachavarria committed:
adding logic to set submittedToLive when submittedToEnvironment is null from api

aachavarria committed:
Publish Dialog Dynamic Warning Messages
Updating createPresenceTable
adding Alert to PublishForm

aachavarria committed:
Fixing space

aachavarria committed:
Cleaning RecentlyPublishedDashlet;
Removing checkboxes on RecentlyPublishedDashlet;

aachavarria committed:
Optimizing imports

aachavarria committed:
Updating RecentActivityDashlet.tsx UI,
Adding Actions bar to RecentActivityDashlet.tsx
Creating RecentActivityDashletGrid Folder

aachavarria committed:
Cleaning Dashboard Code

aachavarria committed:
adding loading to ActionsBar.tsx

aachavarria committed:
Adding type

aachavarria committed:
Adding publishingDate and PublishingTarget to PublishDialog items

rart commented on ui/app/src/components/ApprovedScheduledDashlet/ApprovedScheduledDashlet.tsx:
```diff
@@ -182,11 +191,48 @@ export default function ApprovedScheduledDashlet(props: ApprovedScheduledDashlet
+    if (isAllChecked) {
+      setSelectedLookup({});
+    } else {
+      setSelectedLookup({ ...selectedLookup, ...createPresenceTable(Object.keys(state.itemsLookup), true) });
```
```suggestion
      setSelectedLookup({ ...selectedLookup, ...createPresenceTable(Object.keys(state.itemsLookup)) });
```

rart commented on ui/app/src/components/ApprovedScheduledDashlet/ApprovedScheduledDashlet.tsx:
```diff
@@ -182,11 +191,48 @@ export default function ApprovedScheduledDashlet(props: ApprovedScheduledDashlet
+    const path = item.path;
+    dispatch(completeDetailedItem({ path }));
+    dispatch(
+      showItemMegaMenu({
```
- I think the mega menu completes its item already
- If needed, this should be a single dispatch to batch

rart commented on ui/app/src/components/AwaitingApprovalDashlet/AwaitingApprovalDashlet.tsx:
```diff
@@ -193,11 +200,48 @@ export default function AwaitingApprovalDashlet(props: AwaitingApprovalDashletPr
+    const path = item.path;
+    dispatch(completeDetailedItem({ path }));
+    dispatch(
+      showItemMegaMenu({
```
Same comment as above

rart commented on ui/app/src/components/AwaitingApprovalDashletGrid/AwaitingApprovalDashletGridUI.tsx:
```diff
@@ -163,16 +163,29 @@ export default function AwaitingApprovalDashletGridUI(props: AwaitingApprovalDas
+                                  title={
+                                    itemsLookup[path].live.dateScheduled &&
+                                    asLocalizedDateTime(
+                                      itemsLookup[path].live.dateScheduled,
```
Doesn't seem like you're ever using item.staging instead. Should we?

rart commented on ui/app/src/components/AwaitingApprovalDashletGrid/AwaitingApprovalDashletGridUI.tsx:
```diff
@@ -163,16 +163,29 @@ export default function AwaitingApprovalDashletGridUI(props: AwaitingApprovalDas
-                                    locale.dateTimeFormatOptions
+                                  {itemsLookup[path].live.dateScheduled ? (
+                                    asLocalizedDateTime(
+                                      itemsLookup[path].live.dateScheduled,
```
as above

rart replied:
Not done. Looking directly at `.live`

rart commented on ui/app/src/components/RecentActivityDashlet/RecentActivityDashlet.tsx:
```diff
@@ -178,6 +190,73 @@ export default function RecentActivityDashlet(props: RecentActivityDashletProps)
+    const path = item.path;
+    dispatch(completeDetailedItem({ path }));
+    dispatch(
+      showItemMegaMenu({
```
Same comment as above

rart replied:
not done

rart commented on ui/app/src/components/RecentActivityDashlet/RecentActivityDashlet.tsx:
```diff
@@ -178,6 +190,73 @@ export default function RecentActivityDashlet(props: RecentActivityDashletProps)
+        });
+        return generateMultipleItemOptions(items, formatMessage).filter((option) =>
+          actionsToBeShown.includes(option.id as AllItemActions)
+        );
```
generate mult item options should receive the actions to display

rart commented on ui/app/src/components/RecentlyPublishedDashlet/RecentlyPublishedDashlet.tsx:
```diff
@@ -202,6 +184,22 @@ export default function RecentlyPublishedDashlet(props: RecentlyPublishedWidgetP
+    const path = item.path;
+    dispatch(completeDetailedItem({ path }));
+    dispatch(
+      showItemMegaMenu({
```
Same as above

rart replied:
Not done (batchActions change)

rart commented on ui/app/src/modules/Content/Dependencies/DependencySelection.tsx:
```diff
@@ -206,8 +229,7 @@ export function DependencySelection(props: DependencySelectionProps) {
-                values={{ ellipsis: '&hellip;' }}
+                defaultMessage="Loading Dependencies, please wait..."
               />
             </span>
```
As discussed, remove block and use button with loading state

rart commented on ui/app/src/modules/Content/Dependencies/DependencySelection.tsx:
```diff
@@ -350,7 +373,8 @@ function SelectionList(props: SelectionListProps) {
         <List className={classes.selectionList}>
-          {items.map((item) => {
+          {// @ts-ignore
+          items.map((item) => {
```
Discuss

rart replied:
Why is this resolved and still as is?

rart commented on ui/app/src/modules/Content/Dependencies/DependencySelection.tsx:
```diff
@@ -378,19 +402,65 @@ function SelectionList(props: SelectionListProps) {
-                />
+                <ListItemText id={labelId}>
+                  <Typography variant="subtitle1">{item.label}</Typography>
+                  {item.live && (
```
item.staging not used

rart commented on ui/app/src/utils/hooks/useDetailedItems.ts:
```diff
@@ -0,0 +1,50 @@
+
+export function useDetailedItems(paths: string[]): { itemsByPath: LookupTable<DetailedItem>; isFetching: boolean } {
+  const dispatch = useDispatch();
+  const lookup = useSelection((state) => state.content.itemsBeingFetchedByPath);
```
```suggestion
  const fetchingLookup = useSelection((state) => state.content.itemsBeingFetchedByPath);
```

rart commented on ui/app/src/utils/hooks/useDetailedItems.ts:
```diff
@@ -0,0 +1,50 @@
+    return _items;
+  }, [itemsByPath, paths]);
+  useEffect(() => {
+    let _actions = [];
```
Don't see a collision with the not-underscored name
```suggestion
    let actions = [];
```

rart commented on ui/app/src/modules/Content/Publish/PublishForm.tsx:
```diff
@@ -35,6 +35,7 @@ import TextFieldWithMax from '../../../components/Controls/TextFieldWithMax';
 import GlobalState from '../../../models/GlobalState';
 import FormLabel from '@material-ui/core/FormLabel';
 import { useSelection } from '../../../utils/hooks/useSelection';
+import { Alert } from '@material-ui/lab';
```
```suggestion
import Alert from '@material-ui/lab/Alert';
```

aachavarria committed:
Fixing issues relating with dashlet not using the correct publishDate or scheduleDate

rart commented on ui/app/src/components/RecentlyPublishedDashlet/RecentlyPublishedDashletUI.tsx:
```diff
@@ -149,43 +127,42 @@ export default function RecentlyPublishedDashletUI(props: RecentlyPublishedWidge
+                                <FormattedMessage id="environment.live" defaultMessage="live" />
                               ) : (
-                                <FormattedMessage id="words.staging" defaultMessage="Staging" />
+                                <FormattedMessage id="environment.staging" defaultMessage="staging" />
```
Rollback. Capitalised is fine.
```suggestion
                                <FormattedMessage id="words.staging" defaultMessage="Staging" />
```

rart commented on ui/app/src/components/RecentlyPublishedDashlet/RecentlyPublishedDashletUI.tsx:
```diff
@@ -149,43 +127,42 @@ export default function RecentlyPublishedDashletUI(props: RecentlyPublishedWidge
                             <GlobalAppGridCell className="width20">
                               {itemsLookup[path].stateMap.submittedToLive ? (
-                                <FormattedMessage id="words.live" defaultMessage="Live" />
+                                <FormattedMessage id="environment.live" defaultMessage="live" />
```
Rollback. Capitalised is fine.
```suggestion
                                <FormattedMessage id="words.live" defaultMessage="Live" />
```

rart commented on ui/app/src/modules/Content/Dependencies/DependencySelection.tsx:
```diff
@@ -27,19 +27,26 @@ import ListItemText from '@material-ui/core/ListItemText';
-interface DependencySelectionProps<T extends BaseItem = BaseItem> {
-  items: T[];
+interface DependencySelectionProps {
+  items?: DetailedItem[] | SandboxItem[];
```
Actually, we're only working with detailed items so this should resolve the whole situation.
```suggestion
  items?: DetailedItem[];
```

rart commented on ui/app/src/modules/Content/Dependencies/DependencySelection.tsx:
```diff
@@ -50,10 +57,10 @@ interface DependencySelectionProps<T extends BaseItem = BaseItem> {
   title: any;
   subtitle?: any;
-  items?: T[];
+  items?: DetailedItem[] | SandboxItem[];
```
```suggestion
  items?: DetailedItem[];
```

rart commented on ui/app/src/utils/detailedItem.ts:
```diff
@@ -0,0 +1,37 @@
+
+import { DetailedItem } from '../models/Item';
+
+export function getDateScheduled(item: DetailedItem): string {
```
Still some places not using these utils.

aachavarria committed:
Fixing issues relating with publishedDate

aachavarria committed:
adding empty message to DependencySelection.tsx;
cleaning code on DependencySelection.tsx;
updating PublishDialog.tsx to remove unnecessary code and add loading dependencies button;

aachavarria committed:
Updating empty alert styles

aachavarria committed:
adding includeOnly to generateMultipleItemOptions;

aachavarria committed:
rollback and renames

aachavarria committed:
Using DetailedItem[] on DependencySelection.tsx

aachavarria committed:
Fixing space

rart commented on ui/app/src/utils/hooks/useDetailedItems.ts:
```diff
@@ -0,0 +1,50 @@
+    });
+    if (actions.length) {
+      dispatch(batchActions(actions));
+    }
```
```suggestion
    if (actions.length) {
      dispatch(actions.length > 1 ? batchActions(actions) : actions[0]);
    }
```

rart reviewed (changes requested):
Changes are still pending

aachavarria committed:
removing comment;
fixes on useDetailedItems.ts

aachavarria committed:
adding batchActions

aachavarria committed:
Fixing issue with sort mutating the state

aachavarria committed:
Fixing issues with deleted items

rart commented on ui/app/src/components/ApprovedScheduledDashlet/translations.ts:
```diff
@@ -0,0 +1,38 @@
+  reject: {
+    id: 'dashboardItemsScheduled.reject',
+    defaultMessage: 'Reject'
+  },
```
These are all `words.`

rart commented on ui/app/src/components/AwaitingApprovalDashlet/AwaitingApprovalDashlet.tsx:
```diff
@@ -193,11 +201,50 @@ export default function AwaitingApprovalDashlet(props: AwaitingApprovalDashletPr
+    } else {
+      itemActionDispatcher({
+        site,
+        item: Object.keys(selectedLookup)
+          .filter((path) => selectedLookup[path])
+          .map((path) => state.itemsLookup[path]),
```
Don't understand the filtering of selected lookup against itself? Technically, that filter will always return the full array, no?
```suggestion
        item: Object.keys(selectedLookup)
          .map((path) => state.itemsLookup[path]),
```
Was the filter meant to be against the mapping?
```suggestion
        item: Object.keys(selectedLookup)
          .map((path) => state.itemsLookup[path])
          .filter(Boolean),
```
Still, if the items aren't loaded, they need to be loaded?

rart commented on ui/app/src/components/ApprovedScheduledDashlet/ApprovedScheduledDashlet.tsx:
```diff
@@ -182,11 +193,51 @@ export default function ApprovedScheduledDashlet(props: ApprovedScheduledDashlet
+    } else {
+      itemActionDispatcher({
+        site,
+        item: Object.keys(selectedLookup)
+          .filter((path) => selectedLookup[path])
+          .map((path) => state.itemsLookup[path]),
```
Same as previous comment.

rart commented on ui/app/src/components/AwaitingApprovalDashlet/translations.ts:
```diff
@@ -0,0 +1,34 @@
+  reject: {
+    id: 'awaitingApprovalDashlet.reject',
+    defaultMessage: 'Reject'
+  },
```
`words.`

rart commented on ui/app/src/components/AwaitingApprovalDashlet/translations.ts:
```diff
@@ -0,0 +1,34 @@
+  clear: {
+    id: 'awaitingApprovalDashlet.clear',
+    defaultMessage: 'Clear {count} selected'
+  }
```
perhaps these should all be same key e.g. `common.clearSelectedWithCount`

rart commented on ui/app/src/components/RecentActivityDashlet/translations.ts:
```diff
@@ -0,0 +1,26 @@
+  clear: {
+    id: 'awaitingApprovalDashlet.clear',
+    defaultMessage: 'Clear {count} selected'
+  }
```
As said above (`common.`)

aachavarria committed:
Creating "common.clearSelectedWithCount";
Adding words. to translations;

aachavarria committed:
adding batchActions

aachavarria committed:
adding getDateScheduled util

aachavarria committed:
Fixing space
